authprox/service.go

```diff
@@ -0,0 +1,322 @@
+package authprox
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	bolt "github.com/coreos/bbolt"
+	"github.com/dedis/cothority"
+	"github.com/dedis/kyber"
+	"github.com/dedis/kyber/share"
+	"github.com/dedis/kyber/sign/dss"
+	"github.com/dedis/kyber/suites"
+	"github.com/dedis/onet"
+	"github.com/dedis/onet/log"
+	"github.com/dedis/onet/network"
+)
+
+// ServiceName is the name of the Authentication Proxy service.
+const ServiceName = "AuthProx"
+
+var authProxID onet.ServiceID
+
+type service struct {
+	*onet.ServiceProcessor
+	ctx        context.Context
+	validators map[string]Validator
+	db         *bolt.DB
+	bucket     []byte
+}
+
+func init() {
+	var err error
+	authProxID, err = onet.RegisterNewService(ServiceName, newService)
+	if err != nil {
+		log.ErrFatal(err, "could not register")
+	}
+	network.RegisterMessages(
+		&EnrollRequest{}, &EnrollResponse{},
+		&SignatureRequest{}, &SignatureResponse{},
+		&EnrollmentsRequest{}, &EnrollmentsResponse{}, &EnrollmentInfo{},
+		&ti{}, &dssConfig{},
+	)
+}
+
+// A Validator is able to check the provided authInfo with respect to
+// the thrid-party authentication system. Return the user-id and the extra_data
+// associated with this auth info, or error.
+type Validator interface {
+	FindClaim(issuer string, authInfo []byte) (claim string, extraData string, err error)
+}
+
+func (s *service) registerValidator(t string, v Validator) {
+	if _, ok := s.validators[t]; ok {
+		panic("cannot re-register a validator")
+	}
+	s.validators[t] = v
+}
+
+func (s *service) validator(in string) (Validator, error) {
+	v, ok := s.validators[in]
+	if !ok {
+		return nil, fmt.Errorf("unknown auth type %v", in)
+	}
+	return v, nil
+}
+
+// ti is a struct to encode/decode a pair of type and issuer.
+type ti struct {
+	T, I string
+}
+
+type dssConfig struct {
+	Secret       kyber.Scalar
+	Participants []kyber.Point
+	LongPri      share.PriShare
+	LongPubs     []kyber.Point
+}
+
+func (s *service) find(typ, issuer string) (*dssConfig, error) {
+	k, err := network.Marshal(&ti{T: typ, I: issuer})
```

Comment:
it's best to use `protobuf.Encode`, `network.Marshal` won't produce valid protobuf format because it has an additional header (which we probably don't need)

Reply:
Yeah, I always get confused about this. I converted it towards network.Marshal already once because I remembered this problem, but I remembered it backwards. Fixing.